pkg/insight/insightstore/deployment_store.go

```diff
@@ -0,0 +1,338 @@
+// Copyright 2022 The PipeCD Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package insightstore
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"sort"
+	"time"
+
+	"go.uber.org/zap"
+
+	"github.com/pipe-cd/pipecd/pkg/filestore"
+	"github.com/pipe-cd/pipecd/pkg/insight"
+)
+
+type DeploymentBlockMetadata struct {
+	BlockID       string                     `json:"block_id"`
+	ChunkMetadata []*DeploymentChunkMetadata `json:"chunk_metadata"`
+}
+
+type DeploymentChunkMetadata struct {
+	ChunkID      string `json:"chunk_id"`
+	ChunkIndex   int    `json:"chunk_index"`
+	MinTimestamp int64  `json:"min_timestamp"`
+	MaxTimestamp int64  `json:"max_timestamp"`
+	Count        int    `json:"count"`
+	Completed    bool   `json:"completed"`
+}
+
+type DeploymentChunk struct {
+	ChunkID     string                    `json:"chunk_id"`
+	Deployments []*insight.DeploymentData `json:"deployments"`
+}
+
+func (c *DeploymentChunk) Size() (int64, error) {
+	raw, err := json.Marshal(c)
+	if err != nil {
+		return 0, err
+	}
+
+	return int64(len(raw)), nil
+}
+
+func (c *DeploymentChunk) FindDeployments(from, to int64) []*insight.DeploymentData {
+	out := make([]*insight.DeploymentData, 0, len(c.Deployments))
+	for _, d := range c.Deployments {
+		if from <= d.CompletedAt && d.CompletedAt <= to {
+			out = append(out, d)
+		}
+	}
+	return out
+}
+
+func (m *DeploymentBlockMetadata) FindChunks(from, to int64) []string {
+	var ids []string
+	for _, m := range m.ChunkMetadata {
+		if overlap(from, to, m.MinTimestamp, m.MaxTimestamp) {
+			ids = append(ids, m.ChunkID)
+		}
+	}
+	return ids
+}
+
+func overlap(lhsFrom, lhsTo, rhsFrom, rhsTo int64) bool {
+	return (rhsFrom < lhsTo && lhsFrom < rhsTo)
+}
+
+func (s *store) ListCompletedDeployments(ctx context.Context, projectID string, from, to int64) ([]*insight.DeploymentData, error) {
+	const rangeLimit time.Duration = 2 * 365 * 24 * time.Hour // 2 year
+
+	if from > to {
+		return nil, errInvalidArg
+	}
+	if float64(to-from) > rangeLimit.Seconds() {
+		return nil, errLargeDuration
+	}
+
+	var (
+		fromYear = time.Unix(from, 0).Year()
+		toYear   = time.Unix(to, 0).Year()
+		out      = make([]*insight.DeploymentData, 0)
```

Comment:
@knanao What kind of memory leak can it be?

Reply:
Apparently, the memory allocation has been more optimized since 1.18.
I had misunderstood that any number of cap sizes would ensure twice as many caps.
So, let's keep this implementation.

https://github.com/golang/go/commit/2dda92ff6f9f07eeb110ecbf0fc2d7a0ddd27f9d

Reply:
Thank you. It's hard to predict the capacity in this case so just go with the standard growth formula.